### PR TITLE
[WIP] kubelet: Improved performance of the cadvisor metric collection

### PR DESCRIPTION
**Pull request by bwplotka**
Fixed https://github.com/kubernetes/kubernetes/pull/107960 PR.

As in the previous PR, the vendor is 80% of this PR. Life is hard 🙃 

Most of the work was done by @ehashman 

/cc @ehashman @dgrisonnet 

#### What type of PR is this?

/kind feature
/sig node
/sig instrumentation

#### What this PR does / why we need it:

Significantly improves kubelet perf of cadvisor metric collection on scrape (e.g by Prometheus).

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

Two PRs has to be merged upstream, before moving with this PR:

* [x] https://github.com/prometheus/client_golang/pull/929
* [ ] https://github.com/google/cadvisor/pull/2974

Also:

* [ ] Macro benchmarks are in progress

#### Does this PR introduce a user-facing change?

cAdvisor uses a new type of registry for container metric collection. Users should see significant improvements in kubelet CPU utilization as a result.

### CODE REVIEW

File: pkg/kubelet/server/server.go [hunk 1]
```diff
@@ -369,33 +367,36 @@ func (s *Server) InstallDefaultHandlers() {
...
-	s.addMetricsBucketMatcher("metrics/resource")
 	resourceRegistry := compbasemetrics.NewKubeRegistry()
 	resourceRegistry.CustomMustRegister(collectors.NewResourceMetricsCollector(s.resourceAnalyzer))
 	s.restfulCont.Handle(resourceMetricsPath,
 		compbasemetrics.HandlerFor(resourceRegistry, compbasemetrics.HandlerOpts{ErrorHandling: compbasemetrics.ContinueOnError}),
 	)
 
-	// prober metrics are exposed under a different endpoint
-
-	s.addMetricsBucketMatcher("metrics/probes")
```
Comment (ehashman) [inline pkg/kubelet/server/server.go:398]: Looks like these unrelated metrics buckets were removed @bwplotka, was that intentional? We still need these for non-cadvisor provided metrics.

**Comment by bwplotka** [comment]
Current status:

* There is a discrepancy in cadvisor metric outputs, so added unit test: https://github.com/kubernetes/kubernetes/pull/108277
 
image

* Sorting of output does not work anymore on client_golang, will look into that.
* There is some unexpected performance, but given output is not valid, no point in comparing too much

**Comment by bwplotka** [comment]
Fixed functionality errors, will update soon.

**Comment by bwplotka** [comment]
Current status: 
* Almost all functional problems fixed: one more to go.
* We see 50% of past CPU usage and 10% of past memory allocations. Yet 2x of memory is being kept in heap, so utilization is overall higher. Investigating further. 
* Needs rebase

**Comment by bwplotka** [comment]
Status update:

* After https://github.com/kubernetes/kubernetes/pull/109029 this PR will be much smaller and reasoning about it will be much easier.
* There are two caches in cadvisor with this change. The internal cgroup cache and client-go cache. The second cache allows us to have reduced CPU and allocation, but it keeps extra bytes on top of the internal cgroup cache, that's why we see more bytes in memory being kept. It is unsure if we can get rid of the internal cgroup, since it is not only used by cgroup prometheus metric export, but also for ad-hoc kubelet operations like handling summary ([GetCPUAndMemoryStatS()](https://github.com/bwplotka/kubernetes/blob/0de43974909fc21b125735ddb45cf17e593612ca/pkg/kubelet/server/stats/handler.go#L142)). This is why removing internal cache might have unpredictiable consequences (for us, [Prometheus client-golang](https://github.com/prometheus/client_golang) experts, not cadvisor ones). We are looking with @saswatamcode (in our very limited free time) if we are not missing anything else here, if there are means of easy improvements of this situation without rewriting cadvisor OR alternatives to caching logic we added that solves https://github.com/kubernetes/kubernetes/issues/104459.

If anyone wants to dig into benchmarked profiles I obtained thanks to https://github.com/bwplotka/kubelet-bench setup:

* [Baseline heap profile (23.55 MB)](https://share.polarsignals.com/a437d6e/) vs [This PR heap profile (53 MB)](https://share.polarsignals.com/1543fde/)
* [Baseline CPU profile (6.1 CPU seconds)](https://share.polarsignals.com/a025ffe/) vs [This PR CPU profile (2.6 CPU seconds)](https://share.polarsignals.com/76f11d7/),

**Comment by enj** [comment]
/remove-sig auth

This appears to belong to SIG Node.  Feel free to add back SIG Auth if needed.

**Pull request closed without merging**